libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+#include "soft_timer.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "critical_section.h"
+#include "interrupt.h"
+#include "status.h"
+#include "stm32f0xx_interrupt.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_tim.h"
+
+typedef struct Timer {
+  uint32_t duration;
+  void *context;
+  bool inuse;
+  SoftTimerCallback callback;
+} Timer;
+
+static volatile Timer s_soft_timer_array[SOFT_TIMER_MAX_TIMERS];
+static volatile SoftTimerID s_active_timer_id;
+static volatile uint32_t s_last_update_count;
+static volatile uint32_t s_active_timer_duration;
+
+static void prv_start_timer(uint32_t duration_us) {
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  TIM_Cmd(TIM2, DISABLE);
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+  if (curr_time + duration_us > UINT32_MAX) {
+    TIM_SetCompare1(TIM2, curr_time + duration_us - UINT32_MAX);
+  } else {
+    TIM_SetCompare1(TIM2, curr_time + duration_us);
+  }
+  s_active_timer_duration = duration_us;
+  TIM_Cmd(TIM2, ENABLE);
+}
+
+static uint32_t prv_soft_timer_update_period(void) {
+  uint32_t counter = TIM_GetCounter(TIM2);
+  if (counter - s_last_update_count > 0) {
+    return counter - s_last_update_count;
+  } else {
+    return counter - s_last_update_count + UINT32_MAX;
+  }
+}
+
+static void prv_soft_timer_update(uint32_t time_since_update) {
+  // Default these values to max. If there is ever a newer timer it will replace them. In the event
+  // there are no timers left the defaults will result in the module being reset to a state where no
+  // timers exist.
+  s_last_update_count = TIM_GetCounter(TIM2);
+  SoftTimerID min_time_id = SOFT_TIMER_MAX_TIMERS;
+  uint32_t min_duration = UINT32_MAX;
+
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    // If the timer is in use and wasn't just added then update its duration.
+    if (s_soft_timer_array[i].inuse) {
+      if (time_since_update >= s_soft_timer_array[i].duration) {
+        // For each inuse timer see if the timer expired. If it has run its callback.
+        s_soft_timer_array[i].callback(i, s_soft_timer_array[i].context);
+        s_soft_timer_array[i].inuse = false;
+      } else {
+        // Otherwise, update the duration left on the timer.
+        s_soft_timer_array[i].duration -= time_since_update;
```

Comment:
The other reason I suggested using an always-running hardware timer is so you don't need to update every active timer.